File: blender_extension/operations/material.py
"""Material operations, including Principled BSDF handling."""

from __future__ import annotations

from typing import Any

import bpy

from .. import compatibility, ids
from ..dispatcher import op, read
from ..protocol import BridgeError, ErrorCode, invalid_argument
from . import _common as c
from . import _nodes as n

#: Semantic Principled inputs and how their values are decoded.
PRINCIPLED_SCALARS = (
    "metallic",
    "roughness",
    "ior",
    "alpha",
    "emission_strength",
    "specular",
    "transmission",
    "coat_weight",
    "coat_roughness",
    "sheen_weight",
    "sheen_roughness",
    "anisotropic",
    "anisotropic_rotation",
)
PRINCIPLED_COLORS = ("base_color", "emission_color")


def principled_node(material):
    """The material's Principled BSDF, or ``None``."""
    if not material.use_nodes or material.node_tree is None:
        return None
    output = n.find_output_node(material.node_tree, ("OUTPUT_MATERIAL",))
    if output is not None:
        surface = output.inputs.get("Surface")
        if surface is not None and surface.is_linked:
            candidate = surface.links[0].from_node
            if candidate.type == "BSDF_PRINCIPLED":
                return candidate
    # Fall back to any Principled in the tree: a material still being built up
    # may not be wired to the output yet.
    for node in material.node_tree.nodes:
        if node.type == "BSDF_PRINCIPLED":
            return node
    return None


def require_principled(material):
    node = principled_node(material)
    if node is None:
        raise BridgeError(
            ErrorCode.UNSUPPORTED_OPERATION,
            f"`{material.name}` has no Principled BSDF. Edit its node graph directly with the "
            "shader tools, or recreate the material with `use_nodes: true`.",
            {"material": material.name},
        )
    return node


def read_principled(material) -> dict[str, Any] | None:
    node = principled_node(material)
    if node is None:
        return None
    payload: dict[str, Any] = {}
    for semantic in PRINCIPLED_COLORS:
        socket = compatibility.principled_socket(node, semantic)
        if socket is not None:
            payload[semantic] = c.color_dict(socket.default_value)
    for semantic in PRINCIPLED_SCALARS:
        socket = compatibility.principled_socket(node, semantic)
        if socket is not None:
            try:
                payload[semantic] = float(socket.default_value)
            except TypeError:
                # A socket that is driven by a link has no scalar default worth
                # reporting.
                continue
    normal = compatibility.principled_socket(node, "normal")
    if normal is not None and normal.is_linked:
        source = normal.links[0].from_node
        if source.type == "NORMAL_MAP":
            payload["normal_strength"] = float(source.inputs["Strength"].default_value)
    return payload


def write_principled(material, values: dict) -> list[str]:
    """Apply semantic Principled values. Returns which ones were written."""
    node = require_principled(material)
    written: list[str] = []

    for semantic in PRINCIPLED_COLORS:
        if values.get(semantic) is None:
            continue
        socket = compatibility.require_principled_socket(node, semantic)
        socket.default_value = c.as_color(values[semantic], semantic)
        written.append(semantic)

    for semantic in PRINCIPLED_SCALARS:
        if values.get(semantic) is None:
            continue
        socket = compatibility.require_principled_socket(node, semantic)
        socket.default_value = float(values[semantic])
        written.append(semantic)

    strength = values.get("normal_strength")
    if strength is not None:
        normal = compatibility.principled_socket(node, "normal")
        if normal is None or not normal.is_linked:
            raise BridgeError(
                ErrorCode.UNSUPPORTED_PROPERTY,
                "`normal_strength` needs a normal map connected to the Principled BSDF.",
                {"material": material.name},
            )
        source = normal.links[0].from_node
        if source.type != "NORMAL_MAP":
            raise BridgeError(
                ErrorCode.UNSUPPORTED_PROPERTY,
                f"The Normal input is driven by a {source.bl_idname}, not a Normal Map node, "
                "so there is no strength to set.",
                {"material": material.name, "node_type": source.bl_idname},
            )
        source.inputs["Strength"].default_value = float(strength)
        written.append("normal_strength")

    return written


def apply_settings(material, settings: dict) -> list[str]:
    written: list[str] = []
    blend = settings.get("blend_method")
    if blend is not None:
        # Blender 4.2 replaced `blend_method` with `surface_render_method` for
        # EEVEE Next; older builds still use the old property.
        if hasattr(material, "surface_render_method"):
            material.surface_render_method = "BLENDED" if blend != "OPAQUE" else "DITHERED"
            written.append("surface_render_method")
        elif hasattr(material, "blend_method"):
            material.blend_method = blend
            written.append("blend_method")
        else:
            raise BridgeError(
                ErrorCode.UNSUPPORTED_PROPERTY,
                "This Blender build exposes neither `blend_method` nor `surface_render_method`.",
                {"requested": blend},
            )
    if settings.get("use_backface_culling") is not None:
        material.use_backface_culling = bool(settings["use_backface_culling"])
        written.append("use_backface_culling")
    if settings.get("displacement_method") is not None:
        if hasattr(material, "displacement_method"):
            material.displacement_method = settings["displacement_method"]
            written.append("displacement_method")
    if settings.get("viewport_color") is not None:
        material.diffuse_color = c.as_color(settings["viewport_color"], "viewport_color")
        written.append("viewport_color")
    if settings.get("alpha_threshold") is not None and hasattr(material, "alpha_threshold"):
        material.alpha_threshold = float(settings["alpha_threshold"])
        written.append("alpha_threshold")
    return written


def read_settings(material) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "use_backface_culling": bool(material.use_backface_culling),
        "viewport_color": c.color_dict(material.diffuse_color),
    }
    for attribute, key in (
        ("blend_method", "blend_method"),
        ("surface_render_method", "blend_method"),
        ("displacement_method", "displacement_method"),
        ("alpha_threshold", "alpha_threshold"),
    ):
        if hasattr(material, attribute):
            value = getattr(material, attribute)
            payload[key] = float(value) if isinstance(value, float) else str(value)
    return payload


def summarise(material, *, detail: bool = False) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "id": ids.ensure_id(material),
        "name": material.name,
        "use_nodes": bool(material.use_nodes),
        "users": int(material.users),
        "node_count": len(material.node_tree.nodes) if material.use_nodes and material.node_tree else 0,
    }
    principled = read_principled(material)
    if principled:
        payload["principled"] = principled
    images = _referenced_images(material)
    if images:
        payload["images"] = images
    if detail:
        payload["settings"] = read_settings(material)
        payload["fake_user"] = bool(material.use_fake_user)
    return payload


def _referenced_images(material) -> list[str]:
    if not material.use_nodes or material.node_tree is None:
        return []
    found = []
    for node in material.node_tree.nodes:
        image = getattr(node, "image", None)
        if image is not None:
            found.append(image.name)
    return sorted(set(found))


# --- operations ------------------------------------------------------------


@read("material.list")
def list_materials(ctx, args: dict) -> dict[str, Any]:
    name_filter = c.optional_str(args, "name_contains")
    used_by = c.optional_str(args, "used_by")
    unused = c.optional_bool(args, "unused")

    candidates = list(bpy.data.materials)
    if used_by is not None:
        obj = ids.find_object(used_by)
        names = {slot.material.name for slot in obj.material_slots if slot.material}
        candidates = [m for m in candidates if m.name in names]
    if unused is not None:
        candidates = [m for m in candidates if (m.users == 0) == unused]

    matched = [m for m in candidates if c.matches_name(m.name, name_filter)]
    matched.sort(key=lambda m: m.name)
    window, cursor = c.paginate(matched, args)
    return {
        "materials": [summarise(m) for m in window],
        "total": len(matched),
        "next_cursor": cursor,
        "revision": ctx.revision,
    }


@read("material.get")
def get(ctx, args: dict) -> dict[str, Any]:
    material = c.material_arg(args)
    return {"material": summarise(material, detail=True), "revision": ctx.revision}


@op("material.create")
def create(ctx, args: dict) -> dict[str, Any]:
    name = c.require_str(args, "name")
    use_nodes = c.optional_bool(args, "use_nodes", True)

    material = bpy.data.materials.new(name)
    material.use_nodes = bool(use_nodes)

    principled = c.optional(args, "principled")
    if principled:
        write_principled(material, principled)
    settings = c.optional(args, "settings")
    if settings:
        apply_settings(material, settings)

    assigned = []
    for obj in c.objects_arg(args, "assign_to", required=False):
        slot_index = _assign(obj, material, slot_index=None, replace_all=False)
        assigned.append({"object": ids.ensure_id(obj), "slot_index": slot_index})

    ids.invalidate_cache("material")
    ctx.bump()
    return {
        "material": summarise(material, detail=True),
        "assigned": assigned,
        "revision": ctx.revision,
    }


@op("material.update")
def update(ctx, args: dict) -> dict[str, Any]:
    material = c.material_arg(args)
    changed: list[str] = []

    name = c.optional_str(args, "name")
    if name is not None:
        material.name = name
        changed.append("name")
        ids.invalidate_cache("material")

    principled = c.optional(args, "principled")
    if principled:
        changed.extend(write_principled(material, principled))

    settings = c.optional(args, "settings")
    if settings:
        changed.extend(apply_settings(material, settings))

    if not changed:
        raise invalid_argument("Nothing to update.")

    ctx.bump()
    return {"material": summarise(material, detail=True), "changed": changed, "revision": ctx.revision}


@op("material.duplicate")
def duplicate(ctx, args: dict) -> dict[str, Any]:
    material = c.material_arg(args)
    name = c.optional_str(args, "name")
    copy = material.copy()
    from .. import config

    # A copied data-block inherits the id; a duplicate is a new entity.
    copy.pop(config.ID_PROPERTY, None)
    if name:
        copy.name = name
    ids.invalidate_cache("material")
    ctx.bump()
    return {"material": summarise(copy, detail=True), "revision": ctx.revision}


@op("material.delete")
def delete(ctx, args: dict) -> dict[str, Any]:
    material = c.material_arg(args)
    force = c.optional_bool(args, "force", False)
    if material.users > 0 and not force:
        raise invalid_argument(
            f"`{material.name}` is used by {material.users} data-block(s). "
            "Unassign it first, or pass force:true to delete it anyway.",
            material=material.name,
            users=material.users,
        )
    payload = {"id": ids.ensure_id(material), "name": material.name}
    bpy.data.materials.remove(material)
    ids.invalidate_cache("material")
    ctx.bump()
    return {"deleted": payload, "revision": ctx.revision}


@op("material.assign")
def assign(ctx, args: dict) -> dict[str, Any]:
    material = c.material_arg(args)
    objects = c.objects_arg(args, "objects")
    slot_index = c.optional_int(args, "slot_index")
    replace_all = c.optional_bool(args, "replace_all", False)
    face_indices = [int(i) for i in c.optional_list(args, "face_indices")]
    expected_revision = c.optional_int(args, "expected_mesh_revision")

    results = []
    for obj in objects:
        if face_indices:
            index = _assign_faces(obj, material, face_indices, expected_revision)
        else:
            index = _assign(obj, material, slot_index, replace_all)
        results.append(
            {
                "object_id": ids.ensure_id(obj),
                "material_id": ids.ensure_id(material),
                "slot_index": index,
            }
        )

    ctx.bump()
    return {"assignments": results, "revision": ctx.revision}


def _assign(obj, material, slot_index: int | None, replace_all: bool) -> int:
    if not hasattr(obj.data, "materials"):
        raise BridgeError(
            ErrorCode.UNSUPPORTED_OPERATION,
            f"`{obj.name}` ({obj.type}) cannot hold materials.",
            {"object": obj.name, "type": obj.type},
        )

    if replace_all:
        obj.data.materials.clear()
        obj.data.materials.append(material)
        return 0

    if slot_index is not None:
        if slot_index < 0 or slot_index >= len(obj.material_slots):
            raise invalid_argument(
                f"`{obj.name}` has {len(obj.material_slots)} material slot(s); "
                f"index {slot_index} is out of range.",
                object=obj.name,
                slot_count=len(obj.material_slots),
            )
        obj.material_slots[slot_index].material = material
        return slot_index

    # Reuse an empty slot before adding another: appending blindly leaves
    # objects with a growing list of unused slots.
    for index, slot in enumerate(obj.material_slots):
        if slot.material is None:
            slot.material = material
            return index
        if slot.material == material:
            return index

    obj.data.materials.append(material)
    return len(obj.material_slots) - 1


def _assign_faces(obj, material, face_indices: list[int], expected_revision: int | None) -> int:
    mesh = c.require_mesh(obj)
    ids.check_mesh_revision(mesh, expected_revision)

    slot_index = _assign(obj, material, None, False)
    face_count = len(mesh.polygons)
    out_of_range = [i for i in face_indices if i < 0 or i >= face_count]
    if out_of_range:
        raise invalid_argument(
            f"`{obj.name}` has {face_count} faces; {len(out_of_range)} of the given indices are "
            "out of range.",
            object=obj.name,
            face_count=face_count,
            out_of_range=out_of_range[:20],
        )
    for index in face_indices:
        mesh.polygons[index].material_index = slot_index
    mesh.update()
    return slot_index


@op("material.unassign")
def unassign(ctx, args: dict) -> dict[str, Any]:
    objects = c.objects_arg(args, "objects")
    material = c.material_arg(args, "material", required=False)
    remove_slot = c.optional_bool(args, "remove_slot", True)

    cleared = []
    for obj in objects:
        for index in reversed(range(len(obj.material_slots))):
            slot = obj.material_slots[index]
            if material is not None and slot.material != material:
                continue
            slot.material = None
            cleared.append({"object": ids.ensure_id(obj), "slot_index": index})
            if remove_slot and hasattr(obj.data, "materials") and index < len(obj.data.materials):
                obj.data.materials.pop(index=index)
    ctx.bump()
    return {"cleared": cleared, "revision": ctx.revision}


@read("material.slot.list")
def list_slots(ctx, args: dict) -> dict[str, Any]:
    obj = c.object_arg(args)
    return {
        "object_id": ids.ensure_id(obj),
        "slots": [
            {
                "index": index,
                "material": ids.ensure_id(slot.material) if slot.material else None,
                "name": slot.material.name if slot.material else None,
                "link": slot.link,
            }
            for index, slot in enumerate(obj.material_slots)
        ],
        "revision": ctx.revision,
    }

File: crates/blender-protocol/src/material.rs
//! Material payloads.
//!
//! The Principled BSDF is the practical centre of Blender shading, so it gets a
//! first-class typed surface. Its socket names moved in 4.0 (`Specular` ->
//! `Specular IOR Level`, `Emission` -> `Emission Color`, subsurface reworked),
//! which is exactly why callers set *semantic* fields here and the bridge maps
//! them to whatever the connected build actually exposes.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::{
    BlenderError, Page, Result, Validate, check_name,
    ids::{MaterialId, MaterialRef, ObjectRef},
    math::{Color4, Finite, check_non_negative, check_range},
};

/// Semantic Principled BSDF inputs. Every field is optional; omitted fields are
/// left untouched.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct PrincipledInputs {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub base_color: Option<Color4>,
    /// 0 = dielectric, 1 = metal. Values between are physically meaningless
    /// except as a blend for mixed surfaces.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metallic: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub roughness: Option<f64>,
    /// Index of refraction. 1.45 for glass, 1.33 for water, 1.5 default.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ior: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub alpha: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub emission_color: Option<Color4>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub emission_strength: Option<f64>,
    /// Specular reflection level for dielectrics (`Specular IOR Level` in 4.x).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub specular: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub transmission: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub coat_weight: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub coat_roughness: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sheen_weight: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sheen_roughness: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub anisotropic: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub anisotropic_rotation: Option<f64>,
    /// Strength multiplier for an attached normal map, if there is one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub normal_strength: Option<f64>,
}

impl PrincipledInputs {
    /// `true` when nothing was requested, so callers can skip a round trip.
    pub fn is_empty(&self) -> bool {
        serde_json::to_value(self)
            .map(|v| v.as_object().is_none_or(|o| o.is_empty()))
            .unwrap_or(false)
    }
}

impl Validate for PrincipledInputs {
    fn validate(&self) -> Result<()> {
        for (value, field) in [
            (self.metallic, "metallic"),
            (self.roughness, "roughness"),
            (self.alpha, "alpha"),
            (self.transmission, "transmission"),
            (self.coat_weight, "coat_weight"),
            (self.coat_roughness, "coat_roughness"),
            (self.sheen_weight, "sheen_weight"),
            (self.sheen_roughness, "sheen_roughness"),
            (self.anisotropic, "anisotropic"),
        ] {
            if let Some(v) = value {
                check_range(v, 0.0, 1.0, field)?;
            }
        }
        if let Some(ior) = self.ior {
            // Below 1.0 is not a refractive index; above 3.0 is beyond any
            // material Cycles models usefully.
            check_range(ior, 1.0, 3.0, "ior")?;
        }
        if let Some(specular) = self.specular {
            check_range(specular, 0.0, 2.0, "specular")?;
        }
        for (value, field) in [
            (self.emission_strength, "emission_strength"),
            (self.normal_strength, "normal_strength"),
        ] {
            if let Some(v) = value {
                check_non_negative(v, field)?;
            }
        }
        if let Some(color) = self.base_color {
            color.check_finite("base_color")?;
        }
        if let Some(color) = self.emission_color {
            color.check_finite("emission_color")?;
        }
        if let Some(rotation) = self.anisotropic_rotation {
            check_range(rotation, 0.0, 1.0, "anisotropic_rotation")?;
        }
        Ok(())
    }
}

/// How a material handles transparency in EEVEE and in the viewport.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum BlendMode {
    Opaque,
    Clip,
    Hashed,
    Blend,
}

/// Which side of a face is rendered.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DisplacementMethod {
    Bump,
    Displacement,
    BothDisplacementAndBump,
}

/// Non-shader material settings.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct MaterialSettings {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub blend_method: Option<BlendMode>,
    /// Render both faces of single-sided geometry.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub use_backface_culling: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub displacement_method: Option<DisplacementMethod>,
    /// Viewport solid-mode colour. Purely cosmetic.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub viewport_color: Option<Color4>,
    /// Alpha threshold for `CLIP` blending.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub alpha_threshold: Option<f64>,
}

impl Validate for MaterialSettings {
    fn validate(&self) -> Result<()> {
        if let Some(color) = self.viewport_color {
            color.check_finite("viewport_color")?;
        }
        if let Some(threshold) = self.alpha_threshold {
            check_range(threshold, 0.0, 1.0, "alpha_threshold")?;
        }
        Ok(())
    }
}

/// `material.create`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct CreateMaterial {
    pub name: String,
    /// Start from a Principled BSDF wired to the output. Turn this off to get
    /// an empty node tree to build by hand.
    #[serde(default = "crate::object::default_true")]
    pub use_nodes: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub principled: Option<PrincipledInputs>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub settings: Option<MaterialSettings>,
    /// Objects to assign the new material to immediately.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub assign_to: Vec<ObjectRef>,
}

impl Validate for CreateMaterial {
    fn validate(&self) -> Result<()> {
        check_name(&self.name, "name")?;
        if let Some(principled) = &self.principled {
            if !self.use_nodes {
                return Err(BlenderError::invalid_argument(
                    "`principled` requires `use_nodes: true`; a material without nodes has no Principled BSDF.",
                ));
            }
            principled.validate()?;
        }
        if let Some(settings) = &self.settings {
            settings.validate()?;
        }
        Ok(())
    }
}

/// `material.update`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct UpdateMaterial {
    pub material: MaterialRef,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub principled: Option<PrincipledInputs>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub settings: Option<MaterialSettings>,
}

impl Validate for UpdateMaterial {
    fn validate(&self) -> Result<()> {
        if let Some(name) = &self.name {
            check_name(name, "name")?;
        }
        if let Some(principled) = &self.principled {
            principled.validate()?;
        }
        if let Some(settings) = &self.settings {
            settings.validate()?;
        }
        if self.name.is_none() && self.principled.is_none() && self.settings.is_none() {
            return Err(BlenderError::invalid_argument(
                "`material.update` needs at least one of name, principled or settings.",
            ));
        }
        Ok(())
    }
}

/// `material.assign`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct AssignMaterial {
    pub material: MaterialRef,
    pub objects: Vec<ObjectRef>,
    /// Slot to write into. Omit to append a new slot, or reuse slot 0 when the
    /// object has exactly one empty slot.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub slot_index: Option<u32>,
    /// Replace every existing slot instead of writing one.
    #[serde(default)]
    pub replace_all: bool,
    /// Restrict the assignment to these face indices. Requires exactly one
    /// object, and is subject to the mesh revision check.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub face_indices: Vec<u32>,
    /// Mesh revision the `face_indices` were read at.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expected_mesh_revision: Option<u64>,
}

impl Validate for AssignMaterial {
    fn validate(&self) -> Result<()> {
        if self.objects.is_empty() {
            return Err(BlenderError::invalid_argument(
                "`objects` must name at least one object.",
            ));
        }
        if !self.face_indices.is_empty() && self.objects.len() != 1 {
            return Err(BlenderError::invalid_argument(
                "`face_indices` addresses one mesh, so exactly one object may be given.",
            )
            .with_detail("objects", self.objects.len()));
        }
        if self.replace_all && self.slot_index.is_some() {
            return Err(BlenderError::invalid_argument(
                "`replace_all` and `slot_index` contradict each other.",
            ));
        }
        Ok(())
    }
}

/// `material.list` filters.
#[derive(Debug, Clone, Default, Serialize, Deserialize, JsonSchema)]
pub struct ListMaterials {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name_contains: Option<String>,
    /// Only materials used by this object.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub used_by: Option<ObjectRef>,
    /// Only materials with no users.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub unused: Option<bool>,
    #[serde(default, flatten)]
    pub page: Page,
}

impl Validate for ListMaterials {
    fn validate(&self) -> Result<()> {
        self.page.validate()
    }
}

/// A material as reported by `material.get` / `material.list`.
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct MaterialSummary {
    pub id: MaterialId,
    pub name: String,
    #[serde(default)]
    pub use_nodes: bool,
    #[serde(default)]
    pub users: u32,
    /// Present when the material has a Principled BSDF the bridge could read.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub principled: Option<PrincipledInputs>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub settings: Option<MaterialSettings>,
    #[serde(default)]
    pub node_count: u32,
    /// Image texture files this material references, for missing-texture hunts.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub images: Vec<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn clamps_are_enforced_on_normalised_inputs() {
        let inputs = PrincipledInputs {
            roughness: Some(1.5),
            ..Default::default()
        };
        let err = inputs.validate().unwrap_err();
        assert!(err.message.contains("roughness"), "{}", err.message);
    }

    #[test]
    fn ior_below_one_is_rejected() {
        let inputs = PrincipledInputs {
            ior: Some(0.5),
            ..Default::default()
        };
        assert!(inputs.validate().is_err());
    }

    #[test]
    fn emission_strength_may_exceed_one() {
        let inputs = PrincipledInputs {
            emission_strength: Some(50.0),
            ..Default::default()
        };
        assert!(inputs.validate().is_ok());
    }

    #[test]
    fn principled_without_nodes_is_rejected() {
        let params = CreateMaterial {
            name: "Concrete".into(),
            use_nodes: false,
            principled: Some(PrincipledInputs {
                roughness: Some(0.5),
                ..Default::default()
            }),
            settings: None,
            assign_to: vec![],
        };
        assert!(params.validate().is_err());
    }

    #[test]
    fn face_assignment_requires_a_single_object() {
        let params = AssignMaterial {
            material: MaterialRef::name("Concrete"),
            objects: vec![ObjectRef::name("A"), ObjectRef::name("B")],
            slot_index: None,
            replace_all: false,
            face_indices: vec![1, 2],
            expected_mesh_revision: None,
        };
        assert!(params.validate().is_err());
    }
}

/// `material.get`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GetMaterial {
    pub material: MaterialRef,
}

/// `material.delete`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct DeleteMaterial {
    pub material: MaterialRef,
    /// Delete even though other data-blocks still use it.
    #[serde(default)]
    pub force: bool,
}

/// `material.duplicate`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct DuplicateMaterial {
    pub material: MaterialRef,
    /// Name for the copy. Blender appends a numeric suffix if omitted.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}

/// `material.unassign`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct UnassignMaterial {
    pub objects: Vec<ObjectRef>,
    /// Only clear slots holding this material. Omit to clear every slot.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub material: Option<MaterialRef>,
    /// Remove the emptied slot rather than leaving it blank.
    #[serde(default = "crate::object::default_true")]
    pub remove_slot: bool,
}

/// `material.slot.list`
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ListMaterialSlots {
    pub object: ObjectRef,
}

impl Validate for GetMaterial {}
impl Validate for DeleteMaterial {}
impl Validate for ListMaterialSlots {}

impl Validate for DuplicateMaterial {
    fn validate(&self) -> Result<()> {
        if let Some(name) = &self.name {
            check_name(name, "name")?;
        }
        Ok(())
    }
}

impl Validate for UnassignMaterial {
    fn validate(&self) -> Result<()> {
        if self.objects.is_empty() {
            return Err(BlenderError::invalid_argument(
                "`objects` must name at least one object.",
            ));
        }
        Ok(())
    }
}
